src/include/settings/settings_defs.h

```diff
@@ -185,63 +185,82 @@ SETTING_bool(
 )
 
 SETTING_bool(
-    metrics_logging,
+    use_metrics_thread,
+    "Use a thread for the metrics sub-system (default: true).",
+    true,
+    false,
+    terrier::settings::Callbacks::NoOp
+)
+
+SETTING_bool(
+    logging_metrics_enable,
     "Metrics collection for the Logging component (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsLogging
 )
 
 SETTING_bool(
-    metrics_transaction,
+    transaction_metrics_enable,
     "Metrics collection for the TransactionManager component (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsTransaction
 )
 
 SETTING_bool(
-    metrics_gc,
+    gc_metrics_enable,
     "Metrics collection for the GarbageCollector component (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsGC
 )
 
 SETTING_bool(
-    metrics_query_trace,
+    query_trace_metrics_enable,
     "Metrics collection for Query Traces (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsQueryTrace
 )
 
 SETTING_bool(
-    metrics_execution,
+    execution_metrics_enable,
     "Metrics collection for the Execution component (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsExecution
 )
 
 SETTING_bool(
-    metrics_pipeline,
+    pipeline_metrics_enable,
     "Metrics collection for the ExecutionEngine pipelines (default: false).",
     false,
     true,
     terrier::settings::Callbacks::MetricsPipeline
 )
 
+SETTING_int(
+    pipeline_metrics_interval,
+    "Sampling rate of metrics collection for the ExecutionEngine pipelines with 0 = 100%, 1 = 50%, "
+    "9 = 10%, X = 1/(X+1)% (default: 0 for 100%).",
+    0,
+    0,
+    10,
+    true,
+    terrier::settings::Callbacks::NoOp
```

Comment:
I think you need to add a real callback for this one. Otherwise this only have effects when the system starts, and won't do anything when you change it dynamically. You probably need to add a new function in `MetricsManager` for the callback to call and change the sample interval accordingly.

All of my previous comments on the modeling part are addressed!

Reply:
Right. This makes sense. Added the callback.

Reply:
Thanks! But this is getting a bit tricky now because `Callbacks::MetricsPipeline` will enable the pipeline metrics with 0 sample interval blindly and ignoring the `pipeline_metrics_interval` we set. To get this really working, I think you need to add a new `SetMetricsSampleInterval(const MetricsComponent component, uint32_t sample_interval)` function in `MetricsManager` for the callback to call to change `sample_interval_`. And then add a new `EnableMetric(const MetricsComponent component)` function in `MetricsManager` for `Callbacks::MetricsPipeline` to call, which does not change `sample_interval_` in `MetricsManager`. Make sure you call `SetMetricsSampleInterval` in `DBMain` if `SettingsManager` is enabled, because there can be the case that we do not start the system with pipeline metrics, but enable it later.

I would recommend adding test cases for this in `metrics_test` if you want to add this feature. There's already a `ToggleSettings` test case where you can augment to test the sample interval setting in different scenarios. You can also add a `GetMetricsSampleInterval` function in `MetricsManager` to let you probe the `MetricsManager` state. Also, we happen not to have test for the pipeline metrics in `metrics_test`, so this is a good opportunity to add some now. We don't need to test the actual features being recorded because we already have tests for that, but we can test whether the pipeline metrics are successfully recorded and whether the number of records is correct.